crates/sui-storage/src/mutex_table.rs

```diff
@@ -3,33 +3,80 @@
 
 use std::collections::hash_map::{DefaultHasher, RandomState};
 use std::collections::HashMap;
+use std::error::Error;
+use std::fmt;
 use std::hash::{BuildHasher, Hash, Hasher};
 use std::sync::Arc;
+use std::time::Duration;
 
 use tokio::sync::{Mutex, RwLock};
+use tokio::task::JoinHandle;
 
+type InnerLockTable<K> = HashMap<K, Arc<tokio::sync::Mutex<()>>>;
 // MutexTable supports mutual exclusion on keys such as TransactionDigest or ObjectDigest
 pub struct MutexTable<K: Hash> {
     random_state: RandomState,
-    lock_table: Vec<RwLock<HashMap<K, Arc<tokio::sync::Mutex<()>>>>>,
+    lock_table: Arc<Vec<RwLock<InnerLockTable<K>>>>,
     _k: std::marker::PhantomData<K>,
+    _cleaner: JoinHandle<()>,
 }
 
+#[derive(Debug)]
+pub enum TryAcquireLockError {
+    LockTableLocked,
+    LockEntryLocked,
+}
+
+impl fmt::Display for TryAcquireLockError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "operation would block")
+    }
+}
+
+impl Error for TryAcquireLockError {}
 // Opaque struct to hide tokio::sync::MutexGuard.
 pub struct LockGuard(tokio::sync::OwnedMutexGuard<()>);
 
-impl<'b, K: Hash + std::cmp::Eq + 'b> MutexTable<K> {
-    pub fn new(num_shards: usize, shard_size: usize) -> Self {
-        Self {
-            random_state: RandomState::new(),
-            lock_table: (0..num_shards)
+impl<K: Hash + std::cmp::Eq + Send + Sync + 'static> MutexTable<K> {
+    pub fn new_with_cleanup_duration(
+        num_shards: usize,
+        shard_size: usize,
+        duration: Duration,
+    ) -> Self {
+        let lock_table: Arc<Vec<RwLock<InnerLockTable<K>>>> = Arc::new(
+            (0..num_shards)
                 .into_iter()
                 .map(|_| RwLock::new(HashMap::with_capacity(shard_size)))
                 .collect(),
+        );
+        let cloned = lock_table.clone();
+        Self {
+            random_state: RandomState::new(),
+            lock_table,
             _k: std::marker::PhantomData {},
+            _cleaner: tokio::spawn(async move {
+                loop {
+                    Self::cleanup(cloned.clone());
```

Comment:
I am a little worried that this will loop for ever, preventing clean shutdown. Can we have something that indicates the table is not used any more that can be used to signal it should exit?

Reply:
added a boolean flag which is set to true on `Drop` and breaks the loop